text/0072-kafka-schema-registry.md

```diff
@@ -0,0 +1,123 @@
+
+The goals of the centralized schema repository are to:
+
+- provide an explicit, single source of truth about how data in any given topic should look, and the contract that all consumers and producers of a topic must adhere to
+- provide more stability by enforcing backward compatibility of changes via automation
+- explicitly enumerate ownership of schemas
+- make it easier for consumers to identify and reject bad messages without pausing the whole pipeline
+- provide examples of messages for each schema type, which would help with writing consumer tests
+
+# Supporting Data
+
+Over the last quarter, we have seen many incidents related to schema disagreements at Sentry. Examples include the following: (note: inc numbers are internal to Sentry)
+
+- Post process forwarder (INC-218)
+- Snuba’s transactions consumer (INC-210)
+- Super big consumers (INC-220)
+- Replays consumer (INC-250, INC-281)
+
+Many of these incidents are P1. Since messages are in order, an invalid message often halts consumers and requires manual intervention. This is disruptive to both Sentry engineers and our users (it takes us much longer to recover).
+
+# Options Considered
+
+### **Option A (preferred): Publish a library**
+
+Schemas will be made available as a library for multiple programming languages (Python and Rust to start).
+
+**Client usage example:**
+
+```python
+from sentry_kafka_schemas import get_schema
+
+topic = "events"
+schema = get_schema(topic)
+# => {
+# 	"version": 1,
+# 	"type": "json"
+# 	"compatibility_mode": "backward",
+# 	"schema": {
+# 		"$schema": "http://json-schema.org/draft-07/schema#",
+# 		"type": "object",
+# 		...
+# 	}
+# }
+
+# Get a specific version of the schema
+v1_schema = get_schema(topic, version=1)
+```
+
+**How schemas are stored:**
+
+Topic data will be stored as yaml. For example:
+
+```
+topic: events
+schemas:
+    - version: 1
+    compatibility_mode: backward
+    type: json
+    resource: events.schema.json
+    - version: 2
+    compatibility_mode: none
+    type: avro
+    resource: events_v2.avsc
+```
+
+In this example scenario, we decided to make a change to schemas on the “events” topic. The `events.schema.json` and `events_v2.avsc` files must be present.
+
+**Compatibility modes:**
+
+Each schema version defines it's compatibility mode. There will be 2 to start but more can be added if we want to change or tighten the rules.
+
+- `none` - Any changes are allowed. Generally used if a feature is in dev to allow for fast iteration and breaking changes.
+- `backward` - Allows adding optional fields, removing optional fields, and changing from optional to required and required to optional. Required field cannot be added at once, it must be split into 2 separate releases.
```

Comment:
Why do we consider switching a field from optional to required a backward compatible change? 
I think that should be a breaking change that generate a new version.

Reply:
Fair enough, removed that. Fwiw, I think it's ok if most of our schemas are `none` as I think we do want it to be easy to go from optional to required and vice versa. I think we do want to be able to move quickly even for the more mature datasets.